File: discrete/problems/src/main/java/problems/util/math/Ratio.java
package problems.util.math;

import java.util.Objects;

/**
 * Represents rational number without the loss of precision within integer value bounds.
 */
public final class Ratio extends Number implements Comparable<Ratio> {
  private static final long serialVersionUID = -1002448824652078779L;

  public static final Ratio ZERO = new Ratio(0, 1);

  public static final Ratio ONE = new Ratio(1, 1);

  private final int numerator;
  private final int denominator;

  private Ratio(int numerator, int denominator) {
    this.numerator = numerator;
    this.denominator = denominator;
  }

  public static Ratio of(int numerator, int denominator) {
    if (numerator == 0) {
      return ZERO;
    }

    if (denominator == 0) {
      throw new ArithmeticException("Zero-value denominator is not allowed");
    }

    if (numerator == denominator) {
      return ONE;
    }

    final int g = gcd(numerator, denominator);
    numerator = numerator / g;
    denominator = denominator / g;
    return new Ratio(numerator, denominator);
  }

  public static Ratio fromInteger(int n) {
    return of(n, 1);
  }

  public Ratio mul(Ratio other) {
    return of(numerator * other.numerator,
        denominator * other.denominator);
  }

  public Ratio div(Ratio other) {
    return of(numerator * other.denominator,
        denominator * other.numerator);
  }

  public Ratio add(Ratio other) {
    return of(numerator * other.denominator + other.numerator * denominator,
        denominator * other.denominator);
  }

  public Ratio sub(Ratio other) {
    return new Ratio(numerator * other.denominator - other.numerator * denominator,
        denominator * other.denominator);
  }

  @Override
  public int compareTo(Ratio o) {
    final long self = this.numerator * o.denominator;
    final long other = this.denominator * o.numerator;
    return Long.compare(self, other);
  }

  @Override
  public int intValue() {
    return numerator / denominator;
  }

  @Override
  public long longValue() {
    return intValue();
  }

  @Override
  public float floatValue() {
    return ((float) numerator) / denominator;
  }

  @Override
  public double doubleValue() {
    return ((double) numerator) / denominator;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof Ratio)) return false;
    Ratio ratio = (Ratio) o;
    return numerator == ratio.numerator &&
        denominator == ratio.denominator;
  }

  @Override
  public int hashCode() {
    return Objects.hash(numerator, denominator);
  }

  @Override
  public String toString() {
    if (denominator == 1) {
      return Integer.toString(numerator);
    }
    return Integer.toString(numerator) + '/' + Integer.toString(denominator);
  }

  static int gcd(int a, int b) {
    //return BigInteger.valueOf(a).gcd(BigInteger.valueOf(b)).intValueExact();
    for (int t; b != 0; a = b, b = t % b) {
      t = a;
    }
    return a;
  }
}

File: approx/src/main/kotlin/ratio.kt
package ratio

/**
 * An abstraction over rational number, i.e. number which is represented as a ratio of an integer numerator
 * to integer non-zero denominator.
 */
class Ratio private (val numerator : Int, val denominator : Int) {
  class object {
    val ZERO = Ratio(0, 1)
    val ONE = Ratio(1, 1)

    fun invoke(numerator : Int = 1, denominator : Int = 1) : Ratio {
      if (denominator == 0) {
        throw IllegalArgumentException("denominator can't be zero")
      }
      if (numerator == 0) {
        return ZERO
      }
      if (numerator == denominator) {
        return ONE
      }

      val d = gcd(numerator, denominator)
      return Ratio(numerator / d, denominator / d)
    }

    fun gcd(a : Int, b : Int) : Int {
      if (a == 0) {
        return b
      } else if (b == 0) {
        return a
      } else if (a == b) {
        return a
      } else if (a > b) {
        return gcd(a - b, b)
      } else {
        return gcd(a, b - a)
      }
    }
  }

  // this enables clients to write `a * b` for two ratios
  fun times(other: Ratio) = invoke(this.numerator * other.numerator, this.denominator * other.denominator)

  // a / b
  fun div(other : Ratio) = invoke(this.numerator * other.denominator, this.denominator * other.numerator)

  // a + b
  fun add(other : Ratio) = invoke(this.numerator * other.denominator + this.denominator * other.numerator,
      this.denominator * other.denominator)

  // a - b
  fun sub(other : Ratio) = invoke(this.numerator * other.denominator - this.denominator * other.numerator,
      this.denominator * other.denominator)

  fun copy(numerator: Int? = null, denominator: Int? = null): Ratio = invoke(numerator ?: this.numerator,
      denominator ?: this.denominator)

  override fun equals(other: Any?) : Boolean {
    if (other is Ratio) {
      return this.numerator == other.numerator && this.denominator == other.denominator
    }
    return false
  }

  override fun hashCode() : Int {
    return this.numerator * 31 + this.denominator
  }

  // custom toString(), the automatically generated one would be giving "Ratio(numerator=2, denominator=3)"
  override fun toString() = "$numerator/$denominator"
}

